lib/Canonicalize/EmbedStatementComments.cpp

```diff
@@ -0,0 +1,170 @@
+//
+// This file is distributed under the MIT License. See LICENSE.md for details.
+//
+
+#include "llvm/IR/DebugInfoMetadata.h"
+#include "llvm/IR/DebugLoc.h"
+#include "llvm/IR/Dominators.h"
+#include "llvm/IR/Function.h"
+#include "llvm/IR/IRBuilder.h"
+#include "llvm/Pass.h"
+
+#include "revng/Model/IRHelpers.h"
+#include "revng/Model/LoadModelPass.h"
+#include "revng/PTML/CommentPlacementHelper.h"
+#include "revng/Pipeline/Location.h"
+#include "revng/Pipes/Ranks.h"
+#include "revng/Support/DecompilationHelpers.h"
+
+static std::optional<MetaAddress>
+tryExtractAddress(const llvm::Instruction &I) {
+  if (!I.getDebugLoc() || !I.getDebugLoc()->getScope())
+    return std::nullopt;
+
+  auto DebugLocation = I.getDebugLoc()->getScope()->getName().str();
+  auto Parsed = pipeline::locationFromString(revng::ranks::Instruction,
+                                             DebugLocation);
+  revng_assert(Parsed.has_value());
+
+  MetaAddress Extracted = Parsed->at(revng::ranks::Instruction);
+  revng_assert(Extracted.isValid());
+  return Extracted;
+}
+
+template<typename ResultSet>
+static ResultSet gatherNonStatementAddresses(const llvm::Instruction &I,
+                                             ResultSet &&Result = {}) {
+  if (std::optional MaybeAddress = tryExtractAddress(I))
+    Result.emplace(*MaybeAddress);
+
+  for (const llvm::Use &V : I.operands())
+    if (const llvm::Instruction *Cast = llvm::dyn_cast<llvm::Instruction>(V))
+      if (not isStatement(*Cast))
+        gatherNonStatementAddresses(*Cast, Result);
+
+  return Result;
+}
+
+template<>
+struct yield::StatementTraits<llvm::BasicBlock *> {
+
+  using StatementType = llvm::Instruction *;
+
+  static auto getStatements(llvm::BasicBlock *Node) {
+    return *Node | std::views::filter(isStatement)
+           | std::views::transform([](auto &&R) { return &R; });
+  }
+
+  static std::set<MetaAddress> getAddresses(StatementType Statement) {
+    return gatherNonStatementAddresses<std::set<MetaAddress>>(*Statement);
+  }
+};
+
+static std::string
+addressesToString(range_with_value_type<MetaAddress> auto const &Addresses) {
+  std::string Result = "";
+
+  if (not Addresses.empty()) {
+    for (const MetaAddress &Address : Addresses)
+      Result += Address.toString() + " + ";
+    Result.resize(Result.size() - 3);
+  }
+
+  return Result;
+}
+
+struct EmbedStatementComments : public llvm::FunctionPass {
```

Comment:
Mmmh, you need to do things as described [in this pad](https://pad.rev.ng/s/EVGvDPdZA#).

Reply:
Ah, right, the pipe rework, there was a rev.ng hour about it at some point.

My bad!

Reply:
Done (I went with a pure pipe, so no passes).